### DISCUSSION TITLE
Data Transformation and Mapping

### DISCUSSION THREAD

**Discussion by mgubaidullin** [apache/camel-karavan#693] [Ideas]
We are brainstorming/prototyping the most demanded requirement for Karavan: Data Transformation Mapping.
All aspects: Runtime, Formats, UI/UX, etc.
Share your ideas, thoughts, experience, wishes here 👇 

![mapping-dd](https://user-images.githubusercontent.com/1379213/225989613-ab4aaffa-ef38-47f7-a241-7922bf56ab70.gif)

**Comment by prodejna** [comment]
It would be nice to have operations similar to the data mapper mediator from WSO2.
https://ei.docs.wso2.com/en/latest/micro-integrator/references/mediators/data-Mapper-Mediator/#data-mapper-operations

**Reply by sevEnil** [reply]
I Like the green bullet to show there is a connection like in this video:  https://www.lobster-world.com/wp-content/uploads/2023/05/Mapping-Engine.mp4

**Comment by durgapmenon** [comment]
Getting configuration properties from an externally located file and mapping those configurations to the Kamelets or integration components to external upstream/downstream systems. 

Even though, this is already supported by referencing the properties present in application.properties within the local workspace for a Karavan project, it would be cool if the Data Transformation/Mapping capability could enable this for configuration files that could be located remotely.

**Comment by kshepitko** [comment]
@mgubaidullin a question: are you going to support AtlasMap (I remember some hypothetical plans from your roadmap https://camel.apache.org/blog/2022/10/camel-karavan-roadmap/) , or not? The project looks almost dead, unfortunately. It's quite promising, IMO - but looks abandoned.

Btw, have you checked some alternatives for AtlasMap? Quoting you from the link above:

> We are analyzing existing open-source data mapping solutions like AtlasMap that might be embedded into Karavan Designer.

Have you found something in particular? If so, do you mind sharing your findings? I'm looking for the such solution (ideally backed by a Camel component), but the only mapper with the visual representation I could find was only AtlasMap. Or maybe something without UI, but with the same somewhat universal mapping (XML/JSON -> JSON/XML).

Thanks in advance!

**Comment by guoxunbo** [comment]
Hey. I had archived the AtlasMap in karavan. 
It can transfer xml/json/java -> to xml/jason/java.

**Comment by guoxunbo** [comment]
<img width="1479" alt="image" src="https://user-images.githubusercontent.com/14329149/230303734-a70d7325-64ac-4a98-8a46-1060f235203f.png">
Because altasMap had many error/warning with use.so I create entityMap point to expands altasmap.
<img width="1489" alt="image" src="https://user-images.githubusercontent.com/14329149/230304102-31e041fa-8720-4c89-8e94-0a32565cf7b1.png">
The Mapping view like this. It can use java/json/xml to mapping them later.

**Reply by Praval791** [reply]
I couldn't find `entityMap` in the component or Kamelet. Are you using a custom component? Could you share some resources on how I can create one as well?

**Reply by kondetiramanakumar** [reply]
@guoxunbo  I am also not able to find the entitymap component in the kamelets. please share the resources on how to create the entity map???

**Comment by guoxunbo** [comment]
![Uploading image.png…]()
Here can edit the mapping transformation. such as leftpad, rightpad...

**Comment by vijay31** [comment]
Hello, Is this feature ready ? When can we expect this in Karavan to use ?

**Reply by mgubaidullin** [reply]
This feature is not yet implemented. No due date is planned yet

**Comment by rapetum228** [comment]
Looks very interesting. We are waiting 😊

**Comment by kondetiramanakumar** [comment]
Hi all , As a part of my requirement we required the atlasmap similar feature in the karavan for mapping and transforming  the fields. any suggestions on my requirement???

**Comment by samar-elsayed** [comment]
this looks very promising, I hope you can pick it up and release it 
it will be handy for my team

**Comment by kondetiramanakumar** [comment]
@mgubaidullin 
Hope you are doing great.

I want to know Are there availability of the data transformation and mapping component  in our existing Apache Camel Karavan ???.your swift response would be highly appreciated.

**Comment by AnasAldweik999** [comment]
Hi @mgubaidullin 

I'm currently exploring UI-driven mapping for integrations and came across this discussion. I wanted to ask if there's any update on the current status of the mapping feature mentioned here — is it available for testing or planned for an upcoming release?

Also, is there any documentation or example on how to use it (if it's partially implemented or behind a feature flag)? I'd love to try it out and potentially contribute feedback or ideas.

Thanks in advance
